experiment_prototype/scan_classes/sequences.py

```diff
@@ -104,203 +82,143 @@ def __init__(self, seqn_keys, sequence_slice_dict, sequence_interface, transmit_
                                transmit_metadata)
 
-        # TODO: pass clear frequencies to pass to pulses
 
-        # All interfacing at this point is PULSE.
+        txrate = self.transmit_metadata['txrate']
+        txctrfreq = self.transmit_metadata['txctrfreq']
+        main_antenna_count = self.transmit_metadata['main_antenna_count']
+        main_antenna_spacing = self.transmit_metadata['main_antenna_spacing']
+        pulse_ramp_time = self.transmit_metadata['pulse_ramp_time']
+        max_usrp_dac_amplitude = self.transmit_metadata['max_usrp_dac_amplitude']
+        tr_window_time = self.transmit_metadata['tr_window_time']
+
+        self.basic_slice_pulses = {}
+        single_pulse_timing = []
+        for slice_id in self.slice_ids:
+
+            exp_slice = self.slice_dict[slice_id]
+            tx_freq_khz = float(exp_slice['txfreq'])
+            wave_freq = tx_freq_khz - txctrfreq
+            wave_freq_hz = wave_freq * 1000
+
+            if not exp_slice['rxonly']:
+                basic_samples, real_freq = get_samples(txrate, wave_freq_hz,
+                                                   float(exp_slice['pulse_len']) / 1e6,
+                                                   pulse_ramp_time,
+                                                   1.0,
+                                                   exp_slice['iwavetable'],
+                                                   exp_slice['qwavetable'])
+                if real_freq != wave_freq_hz:
+                    errmsg = 'Actual Frequency {} is Not Equal to Intended Wave Freq {}'.format(real_freq,
+                                                                                                wave_freq_hz)
+                    raise ExperimentException(errmsg)  # TODO change to warning? only happens on non-SINE
+
+                # convert the beam angles to rads
+                beam_rads = np.pi / 180 * np.array(exp_slice['beam_angle'], dtype=np.float64)
+
+                main_antennas = np.arange(main_antenna_count)
+                x = (((main_antenna_count - 1) / 2.0 - main_antennas) * main_antenna_spacing)
+                x *= 2 * np.pi * (tx_freq_khz * 1000)
+
+                y = np.cos(np.pi / 2.0 - beam_rads) / speed_of_light
+
+                # split up the calculations for beams and antennas. Outer multiply of the two
+                # vectors will yield all antenna phases needed for each beam.
+                phase_shift = np.fmod(np.outer(y, x), 2.0 * np.pi) # beams by antenna
+                phase_shift = np.exp(1j * phase_shift)
+
+                # We want to apply all the phases to the basic samples. We can flatten the phases
+                # so that can multiply them all with the basic samples. This can later be reshaped
+                # so that each antenna has a set of phased samples for each beam.
+                phased_samps_for_beams = np.outer(phase_shift.flatten(), basic_samples)
```

Comment:
Yeah, essentially you want a set of samples for every beam phase of every antenna. Because you are multiplying the same base of samples to all beam phases, then you can flatten, multiply and then reshape. There are other ways to do it, but this is very easy.